src/stream/src/executor/backfill.rs

```diff
@@ -0,0 +1,380 @@
+                }
+            };
+
+            tracing::debug!(
+                actor = self.actor_id,
+                "Backfill has already finished and forward messages directly to the downstream"
+            );
+
+            // Backfill has already finished.
+            // Forward messages directly to the downstream.
+            let upstream = upstream
+                .map(move |result| result.map(|msg| Self::mapping_message(msg, &upstream_indices)));
+            #[for_await]
+            for msg in upstream {
+                let msg: Message = msg?;
+                finish_on_barrier(&msg);
+                yield msg;
+            }
+        } else {
+            // Forward messages directly to the downstream.
+            let upstream = upstream
+                .map(move |result| result.map(|msg| Self::mapping_message(msg, &upstream_indices)));
+            #[for_await]
+            for message in upstream {
+                yield message?;
+            }
+        }
+    }
+
+    #[expect(clippy::needless_lifetimes, reason = "code generated by try_stream")]
+    #[try_stream(ok = Option<StreamChunk>, error = StreamExecutorError)]
+    async fn snapshot_read(table: &StorageTable<S>, epoch: u64, current_pos: Option<Row>) {
+        // `current_pos` is None means it needs to scan from the beginning, so we use Unbounded to
+        // scan. Otherwise, use Excluded.
+        let range_bounds = if current_pos.is_none() {
+            (Bound::Unbounded, Bound::Unbounded)
+        } else {
+            (Bound::Excluded(current_pos.unwrap()), Bound::Unbounded)
+        };
+        let iter = table
+            .batch_iter_with_pk_bounds(
+                HummockReadEpoch::Committed(epoch),
+                Row::empty(),
+                range_bounds,
+            )
+            .await?;
+
+        pin_mut!(iter);
+
+        while let Some(data_chunk) = iter
+            .collect_data_chunk(table.schema(), Some(1024))
+            .stack_trace("backfill_snapshot_read")
+            .await?
+        {
+            if data_chunk.cardinality() != 0 {
+                let ops = vec![Op::Insert; data_chunk.capacity()];
+                let stream_chunk = StreamChunk::from_parts(ops, data_chunk);
+                yield Some(stream_chunk);
+            }
+        }
+
+        yield None;
+    }
+
+    /// Mark chunk:
+    /// For each row of the chunk, forward it to downstream if its pk <= `current_pos`, otherwise
+    /// ignore it. We implement it by changing the visibility bitmap.
+    fn mark_chunk(
+        chunk: StreamChunk,
+        current_pos: &Row,
+        table_pk_indices: &PkIndices,
+    ) -> StreamChunk {
+        let chunk = chunk.compact();
+        let (data, ops) = chunk.into_parts();
+        let mut new_visibility = BitmapBuilder::with_capacity(ops.len());
+        for v in data
+            .rows()
+            .map(|row| &row.row_by_indices(table_pk_indices) <= current_pos)
```

Comment:
The performance is bad here due to the allocation of every row. 😢 We may improve this after #6404.